### PR TITLE
Add multinode tests by simulating multiple nodes using Docker.

### PR DESCRIPTION
**Pull request by jssmith**
For testing we want to be able to simulate a Ray cluster using a number of Docker instances running on a single host. This change adds scripts to boot a cluster matching a specific configuration and to run a test script on that cluster.

Other requirements:
- Test scripts should run from the command line as well as from within the cluster environment
- One continuous integration host should simultaneously run tests from different Ray revisions

### CODE REVIEW

**Comment by robertnishihara** [comment]
After doing `build-docker.sh`, the various tests can be run from the command line with something like

```
python $ROOT_DIR/multi_node_docker_test.py \
    --num-nodes=5 \
    --test-script=/ray/test/jenkins_tests/multi_node_tests/test_0.py
    --development-mode
```

The `--development-mode` flag copies the local test files into the head node's docker container so that you can edit the tests and run them without rebuilding the docker image.

**Comment by robertnishihara** [comment]
After this PR is merged, all PR's should be required to pass the Travis tests as well as the Jenkins tests (the multi-node docker stuff in this PR is only being run on Jenkins).

**Comment by robertnishihara** [comment]
The script `./test/jenkins_tests/run_multi_node_tests.sh` is what is running on jenkins.

File: test/jenkins_tests/multi_node_docker_test.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+  parser.add_argument("--test-script", required=True, help="test script")
+  parser.add_argument("--development-mode", action="store_true",
+                      help="use local copies of the test scripts")
+  args = parser.parse_args()
+
+  d = DockerRunner()
+  d.start_ray(mem_size=args.mem_size, shm_size=args.shm_size,
+              num_nodes=args.num_nodes, docker_image=args.docker_image,
+              development_mode=args.development_mode)
+  time.sleep(2)
```
Comment (jssmith) [inline test/jenkins_tests/multi_node_docker_test.py]: Any ideas on a better way to know that Ray has started?

Comment (robertnishihara) [inline test/jenkins_tests/multi_node_docker_test.py]: Good point, now that I think about it, it should be possible to just go ahead and run the test script even if Ray hasn't started (if Ray hasn't started yet, then the call to `ray.init` should just retry as needed).

Comment (robertnishihara) [inline test/jenkins_tests/multi_node_docker_test.py]: I just tried it out, it works like I said with the caveat that `ray.init` only retries for a few seconds and then raises an exception.

**Pull request merged**